pkg/apis/seedmanagement/v1alpha1/types_managedseed.go

```diff
@@ -0,0 +1,216 @@
+// Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1alpha1
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	gardenletconfigv1alpha1 "github.com/gardener/gardener/pkg/gardenlet/apis/config/v1alpha1"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ManagedSeed represents a Shoot that is registered as Seed.
+type ManagedSeed struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard object metadata.
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+	// Specification of the ManagedSeed.
+	Spec ManagedSeedSpec `json:"spec,omitempty" protobuf:"bytes,2,opt,name=spec"`
+	// Most recently observed status of the ManagedSeed.
+	Status ManagedSeedStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`
```

Comment:
```suggestion
	// Specification of the ManagedSeed.
	// +optional
	Spec ManagedSeedSpec `json:"spec,omitempty" protobuf:"bytes,2,opt,name=spec"`
	// Most recently observed status of the ManagedSeed.
	// +optional
	Status ManagedSeedStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`
```

According to the API conventions, if the `omitempty` tag is set, `+optional` should also be added:

> optional fields should also have the `omitempty` struct tag

I addition to that, I am also confused by some types in our core API and also some k8s APIs, that have `optional` and `omitempty` tags for `Shoot.Spec` or `StatefulSet.Spec`, although they are not of a pointer type. According to the conventions, optional fields should be of a pointer type.

So, I'm actually not sure, what is the correct way to do this.

Reply:
Let's do nothing here and keep it as it is. I think spec/status are kind of special because they are first-class citizens of quasi every object. Without these tags `kubectl` would complain too much, I think. Probably that's also the main motivation for the k/k folks and us to NOT follow the conventions in this special case with spec/status.

Reply:
Ok. Should we then also add the `+optional` tag for `ManagedSeed.{Spec,Status}` to keep it consistent with our other APIs and the k8s APIs?

Reply:
I understand about `Status`, but is `Spec` really optional? Yes, I see it's present in k8s APIs, but do you understand why it actually makes sense?

Reply:
I will add the `+optional` tags. BTW there are a few other gardener types where these tags are missing on `Status` and `Spec`, e.g. `Seed`.